File: src/domain/buildDownloadJobs.ts
import {
  availableMonths,
  availableYears,
  isKnownCatalog,
  monthSelectionKey,
  type CatalogCode,
} from "./datasusCatalog";
import type { SelectionState } from "./selectionModel";

export type DownloadJob = {
  catalog: CatalogCode;
  year: number;
  month: number;
  key: string;
};

export type InvalidSelection = {
  level: "catalog" | "year" | "month";
  value: string;
  reason: string;
};

export type BuildDownloadJobsResult = {
  jobs: DownloadJob[];
  invalid: InvalidSelection[];
};

export function buildDownloadJobs(selection: SelectionState): BuildDownloadJobsResult {
  const validYears = new Set(availableYears());
  const jobMap = new Map<string, DownloadJob>();
  const invalid: InvalidSelection[] = [];

  const addJob = (catalog: CatalogCode, year: number, month: number): void => {
    const key = monthSelectionKey(catalog, year, month);
    jobMap.set(key, { catalog, year, month, key });
  };

  for (const catalog of selection.catalogsAll) {
    if (!isKnownCatalog(catalog)) {
      invalid.push({ level: "catalog", value: catalog, reason: "unknown catalog" });
      continue;
    }
    for (const year of validYears) {
      for (const month of availableMonths()) {
        addJob(catalog, year, month);
      }
    }
  }

  for (const yearKey of selection.yearsAll) {
    const parsed = parseYearKey(yearKey);
    if (!parsed.ok) {
      invalid.push({ level: "year", value: yearKey, reason: parsed.reason });
      continue;
    }
    const { catalog, year } = parsed;
    if (!isKnownCatalog(catalog)) {
      invalid.push({ level: "year", value: yearKey, reason: "unknown catalog" });
      continue;
    }
    if (!validYears.has(year)) {
      invalid.push({ level: "year", value: yearKey, reason: "year out of range" });
      continue;
    }
    for (const month of availableMonths()) {
      addJob(catalog, year, month);
    }
  }

  for (const monthKey of selection.months) {
    const parsed = parseMonthKey(monthKey);
    if (!parsed.ok) {
      invalid.push({ level: "month", value: monthKey, reason: parsed.reason });
      continue;
    }
    const { catalog, year, month } = parsed;
    if (!isKnownCatalog(catalog)) {
      invalid.push({ level: "month", value: monthKey, reason: "unknown catalog" });
      continue;
    }
    if (!validYears.has(year)) {
      invalid.push({ level: "month", value: monthKey, reason: "year out of range" });
      continue;
    }
    if (month < 1 || month > 12) {
      invalid.push({ level: "month", value: monthKey, reason: "month out of range" });
      continue;
    }
    addJob(catalog, year, month);
  }

  const jobs = [...jobMap.values()].sort((a, b) => {
    if (a.catalog !== b.catalog) return a.catalog.localeCompare(b.catalog);
    if (a.year !== b.year) return a.year - b.year;
    return a.month - b.month;
  });

  return { jobs, invalid };
}

function parseYearKey(value: string):
  | { ok: true; catalog: string; year: number }
  | { ok: false; reason: string } {
  const parts = value.split(":");
  if (parts.length !== 2) return { ok: false, reason: "invalid key format" };
  const year = Number(parts[1]);
  if (!Number.isInteger(year)) return { ok: false, reason: "invalid year" };
  return { ok: true, catalog: parts[0], year };
}

function parseMonthKey(value: string):
  | { ok: true; catalog: string; year: number; month: number }
  | { ok: false; reason: string } {
  const parts = value.split(":");
  if (parts.length !== 3) return { ok: false, reason: "invalid key format" };
  const year = Number(parts[1]);
  const month = Number(parts[2]);
  if (!Number.isInteger(year)) return { ok: false, reason: "invalid year" };
  if (!Number.isInteger(month)) return { ok: false, reason: "invalid month" };
  return { ok: true, catalog: parts[0], year, month };
}

File: internal/domain/build_download_jobs.go
package domain

import (
	"fmt"
	"sort"
	"strconv"
	"strings"
)

type DownloadJob struct {
	Catalog CatalogCode `json:"catalog"`
	Year    int         `json:"year"`
	Month   int         `json:"month"`
	Key     string      `json:"key"`
}

type InvalidSelection struct {
	Level  string `json:"level"`
	Value  string `json:"value"`
	Reason string `json:"reason"`
}

type BuildDownloadJobsResult struct {
	Jobs    []DownloadJob      `json:"jobs"`
	Invalid []InvalidSelection `json:"invalid,omitempty"`
}

func BuildDownloadJobs(selection SelectionIntent) BuildDownloadJobsResult {
	normalized := selection.Normalize()
	validYears := make(map[int]struct{})
	for _, year := range AvailableYears() {
		validYears[year] = struct{}{}
	}

	jobMap := map[string]DownloadJob{}
	invalid := make([]InvalidSelection, 0)

	addJob := func(catalog CatalogCode, year int, month int) {
		key := MonthSelectionKey(catalog, year, month)
		jobMap[key] = DownloadJob{
			Catalog: catalog,
			Year:    year,
			Month:   month,
			Key:     key,
		}
	}

	for _, catalog := range normalized.CatalogsAll {
		if !IsKnownCatalog(catalog) {
			invalid = append(invalid, InvalidSelection{Level: "catalog", Value: string(catalog), Reason: "unknown catalog"})
			continue
		}
		for year := range validYears {
			for month := 1; month <= 12; month++ {
				addJob(catalog, year, month)
			}
		}
	}

	for _, yearSelectionKey := range normalized.YearsAll {
		catalog, year, err := parseYearSelection(yearSelectionKey)
		if err != nil {
			invalid = append(invalid, InvalidSelection{Level: "year", Value: yearSelectionKey, Reason: err.Error()})
			continue
		}
		if !IsKnownCatalog(catalog) {
			invalid = append(invalid, InvalidSelection{Level: "year", Value: yearSelectionKey, Reason: "unknown catalog"})
			continue
		}
		if _, ok := validYears[year]; !ok {
			invalid = append(invalid, InvalidSelection{Level: "year", Value: yearSelectionKey, Reason: "year out of range"})
			continue
		}
		for month := 1; month <= 12; month++ {
			addJob(catalog, year, month)
		}
	}

	for _, monthSelectionKey := range normalized.Months {
		catalog, year, month, err := parseMonthSelection(monthSelectionKey)
		if err != nil {
			invalid = append(invalid, InvalidSelection{Level: "month", Value: monthSelectionKey, Reason: err.Error()})
			continue
		}
		if !IsKnownCatalog(catalog) {
			invalid = append(invalid, InvalidSelection{Level: "month", Value: monthSelectionKey, Reason: "unknown catalog"})
			continue
		}
		if _, ok := validYears[year]; !ok {
			invalid = append(invalid, InvalidSelection{Level: "month", Value: monthSelectionKey, Reason: "year out of range"})
			continue
		}
		if month < 1 || month > 12 {
			invalid = append(invalid, InvalidSelection{Level: "month", Value: monthSelectionKey, Reason: "month out of range"})
			continue
		}
		addJob(catalog, year, month)
	}

	jobs := make([]DownloadJob, 0, len(jobMap))
	for _, job := range jobMap {
		jobs = append(jobs, job)
	}
	sort.Slice(jobs, func(i, j int) bool {
		if jobs[i].Catalog != jobs[j].Catalog {
			return jobs[i].Catalog < jobs[j].Catalog
		}
		if jobs[i].Year != jobs[j].Year {
			return jobs[i].Year < jobs[j].Year
		}
		return jobs[i].Month < jobs[j].Month
	})

	return BuildDownloadJobsResult{
		Jobs:    jobs,
		Invalid: invalid,
	}
}

func parseYearSelection(key string) (CatalogCode, int, error) {
	parts := strings.Split(key, ":")
	if len(parts) != 2 {
		return "", 0, fmt.Errorf("invalid key format")
	}
	year, err := strconv.Atoi(parts[1])
	if err != nil {
		return "", 0, fmt.Errorf("invalid year")
	}
	return CatalogCode(parts[0]), year, nil
}

func parseMonthSelection(key string) (CatalogCode, int, int, error) {
	parts := strings.Split(key, ":")
	if len(parts) != 3 {
		return "", 0, 0, fmt.Errorf("invalid key format")
	}
	year, err := strconv.Atoi(parts[1])
	if err != nil {
		return "", 0, 0, fmt.Errorf("invalid year")
	}
	month, err := strconv.Atoi(parts[2])
	if err != nil {
		return "", 0, 0, fmt.Errorf("invalid month")
	}
	return CatalogCode(parts[0]), year, month, nil
}
